xarray/core/missing.py

```diff
@@ -0,0 +1,213 @@
+                 copy=False, bounds_error=False, **kwargs):
+        from scipy.interpolate import interp1d
+
+        if kind is None:
+            raise ValueError('kind is a required argument')
+
+        self.cons_kwargs = kwargs
+        self.call_kwargs = {}
+
+        if fill_value is None and kind == 'linear':
+            fill_value = kwargs.pop('fill_value', (np.nan, yi[-1]))
+        else:
+            fill_value = np.nan
+
+        self.f = interp1d(xi, yi, kind=self.kind, fill_value=fill_value,
+                          **self.cons_kwargs)
+
+
+class FromDerivativesInterpolator(BaseInterpolator):
+    def __init__(self, xi, yi, kind=None, fill_value=None, **kwargs):
+        from scipy.interpolate import BPoly
+
+        self.kind = kind
+        self.cons_kwargs = kwargs
+
+        if fill_value is not None:
+            raise ValueError('from_derivatives does not support fill_value')
+
+        self.f = BPoly.from_derivatives(xi, yi, **self.cons_kwargs)
+
+
+def get_clean_interp_args(arr, method, dim, **kwargs):
+    '''normalize interolation arguments and choose which index to use
+
+    provides compatability with pandas interolation methods'''
+    axis = arr.get_axis_num(dim)
+    if method == 'linear':
+        index = np.arange(arr.shape[axis])
+    elif method in ['values', 'index']:
+        index = arr.get_index(dim)
+        method = 'linear'
+    elif method == 'time':
+        index = arr.get_index(dim)
+        method = 'linear'
+        assert isinstance(index, pd.DatetimeIndex)
+    elif method == 'polynomial':
+        order = kwargs.pop('order', None)
+        if not order:
+            raise ValueError("order needs to be specified and greater than 0")
+        method = order
+        index = np.arange(arr.shape[axis])
+    else:
+        index = arr.get_index(dim)
+
+    if isinstance(index, pd.DatetimeIndex):
+        index = index.values.astype(np.float64)
+
+    # check index sorting now so we can skip it later
+    if not (np.diff(index) > 0).all():
+        raise ValueError("Index must be monotonicly increasing")
+    return method, index
+
+
+def interp_na(self, dim=None, method='linear', inplace=False, **kwargs):
+    '''Interpolate values according to different methods.'''
+
+    arr = self if inplace else self.copy()
+
+    if dim is None:
+        raise NotImplementedError('dim is a required')
+
+    method, index = get_clean_interp_args(arr, method, dim, **kwargs)
+
+    try:
+        interpolator = interpolators[method]
+    except:
+        raise ValueError('%r is not a valid interpolator' % method)
+
+    return apply_ufunc(wrap_interpolator,
+                       interpolator, index, arr, kwargs=kwargs,
+                       input_core_dims=[[], [dim], [dim]],
+                       output_core_dims=[[dim]],
+                       dask='forbidden',
+                       vectorize=True,
+                       keep_attrs=True).transpose(*arr.dims)
```

Comment:
@shoyer - can you take a look at this? About half of my tests are passing but I suspect I'm still missing something in the `apply_ufunc` call. 

I'm getting a few errors that look like this: 

```
ValueError: inconsistent size for core dimension 'dim0': 1 vs 6
```

Reply:
I think the problem here is that you are treating your `interpolator` as an argument to vectorize over. This might technically work (by interpreting `interpolator` as a scalar), but was certainly was not intended.

Possibly we should have a more informative error message here.

There are various ways you can fix this (using either a class or function returned by a function), but either way probably want the way you use `apply_ufunc` to be something more like `apply_ufunc(wrap_interpolator(interpolator), index, arr, ...)`.